src/pnpq/devices/odl_ozoptics_650ml.py

```diff
@@ -4,63 +4,135 @@
 import serial
 from serial import Serial
 from pnpq.devices.optical_delay_line import OpticalDelayLine
-import time
+import time, logging
+from pnpq.errors import (
+    DeviceDisconnectedError,
+    OdlMoveNotCompleted,
+    OdlHomeNotCompleted,
+    OdlGetPosNotCompleted,
+    OdlMoveOutofRangeError,
+)
+
 
 
 class OdlOzOptics(OpticalDelayLine):
+    max_move: int
+    """maximum length for moving odl from 0-position, the total path length will be 2x(max_move)"""
+
+    min_move: int
+    """minimum specified length for moving: usually zero"""
+
+    resolution: int
+    """represents minimum specified length for moving on the stage: usually zero"""
+
+    home_timeout: int
+    """required time for completing home in seconds"""
+
+    move_timeout: int
+    """required time for completing home in seconds"""
+
     def __init__(
         self,
         serial_port: str | None = None,
         serial_number: str | None = None,
         config_file=None,
     ):
         super().__init__(serial_port, serial_number, config_file)
+        self.name = "OzOptics ODL"
+
         self.conn.baudrate = 9600
         """Basic Communication BaudRate"""
         self.resolution = 32768 / 5.08
         """32768 steps per motor revolution(5.08 mm = 2xDistance Travel or mirror travel per pitch 0.1 inch)"""
-
+        self.logger = logging.getLogger(f"{self}")
         self.command_terminate = "\r\n"
+        self.max_move = 50
+        self.min_move = 0
+        self.logger = logging.getLogger(f"{self}")
 
         try:
             self.conn.open()
+
         except:
             raise RuntimeError("Can not open OZ optic ODL device")
 
-    def connect(self):
-        if self.conn.is_open == 0:
+    def __ensure_port_open(self) -> None:
+        if not self.conn.is_open:
+            self.logger.error("disconnected")
+            raise DeviceDisconnectedError(f"{self} is disconnected")
+
+    def reconnect(self):
+        if self.conn.is_open:
+            self.logger.warning(f"Serial connectio to ODL({self}) is already open!")
+        else:
+            self.logger.info(f"Openning connection to ODL({self}) device!")
             try:
                 self.conn.open()
-            except Exception as err:
-                raise RuntimeError("Connection failed: " + str(err))
+            except:
+                raise DeviceDisconnectedError(f"{self} is disconnected")
 
     def move(self, dist: float):
-        if not self.conn.is_open:
-            raise RuntimeError("Moving ODL failed: can not connect to ODL device")
-        if dist > 200 or dist < 0:
-            raise Exception("Invalid Move Parameter")
+        self.logger.debug(f"ODL({self}): move({dist}) command has been received")
```

Comment:
"has been received" is confusing. Actually, the command has not been received by the ODL yet. In fact, we haven't even sent it yet.

Writing method entry/exit logs manually is not a good idea. The best thing to do would be to [create a decorator](https://duckduckgo.com/?t=ffab&q=python+logging+decorators&ia=web).

But even if we don't want to use a decorator, there are ways to improve this.

For example, if the ODL should be identified in every log line, we could find a way to use logging context and logging filters to add that information to every log line, instead of manually adding it to each line.

But, for now, instead of implementing anything fancy like decorators and logging contexts, **I think the best way to solve this problem is not to log method entry/exit.** Instead, log the commands in `serial_command`. This is much simpler, and gives you almost as much information.

Reply:
Actually this is the indication of receiving a command (such as move or home) from the host controller point of view. The we translate the command to ODL-device sequence command for example "S+{something}" and "S?" 
 or "GR". 
In this step I'll log the sequence commands in ` serial_command `. For future I'll create the decorator!

Reply:
I don't know what you mean by "host controller."

This log line marks entry into a function.

If "host controller" means something like `testbed-node`, that code is not part of the PnPQ library. Describing testbed-node actions inside of PnPQ breaks [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)).